### PR TITLE
spec: draft

### PR DESCRIPTION
**Pull request by CosmicToast**
First specification draft.
Note that the specification is not the source of truth, and is purely here to help people implementing CNI understand it faster without learning raku.

### CODE REVIEW

**Comment by CosmicToast** [comment]
Open for comments.
CC @Johann150

File: SPEC.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
```
Comment (Johann150) [inline SPEC.md:40]: So is trailing (non-vertical) white space part of the value? If so, there should be a test for that.
And what happens with he how does this interact with the flexspace extension, is it still part of the value?

File: SPEC.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
+If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.
+
+Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
```
Comment (Johann150) [inline SPEC.md]:
```suggestion
A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next vertical space.
```

**Comment by formerly-a-trickster** [comment]
I'd like to have a well-defined resolution strategy for files that use the same key several times.

Like for example, the last definition of that key wins. So the following

```
source = main.zip
source = src.zip
```

would result in `source=src.zip`

File: SPEC.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.
+
+## Core Language
+A CNI document consists of an arbitrary amount of statements.
+A statement may either be a section declaration, key-value pair, or key-rawvalue pair.
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
```
Comment (Johann150) [inline SPEC.md]: If you would add something about the semantic meaning of the dots, the second sentence would make more sense.

File: SPEC.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
+If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.
+
+Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
```
Comment (Johann150) [inline SPEC.md]: Since values may not contain `#`, starting "inside of a value" is technically impossible. Maybe you have a better idea how to formulate this, mine would be:
```suggestion
A comment may start on its own line, at any point "inside" of a value (thus ending it), after a section declaration, or after a raw value.
```

File: SPEC.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
```
Comment (Johann150) [inline SPEC.md]:
```suggestion
For compatibility with INI, the `;` character may also be used for comments, i.e. it is treated as if it were a `#`, except in raw-values.
```

File: SPEC.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
+
+### Flexspace
+In implementations with the "flexspace" extension, whitespace within statements is not significant.
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
+
```
Comment (Johann150) [inline SPEC.md]:
```suggestion
This also means that comments may be placed inside of section headings or between keys and the `=`, if the construct's syntax is continued on the next line.

```
The phrasing is not the best with this one, but I hope you get the point.

File: SPEC.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
```
Comment (Johann150) [inline SPEC.md]: This is tested in `tests/core/bareword/03.cni` too. If this ability is only added in this extension, that test should be (re)moved.

File: SPEC.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+Backend storage should be implemented in a trie-like fashion, separated by `.`s.
+It should be possible to get the direct list of children of a trie node, as well as the full list of children of a trie node.
+It should be possible to know if a given key is a section, key, both, or neither.
+It should be possible to walk over matches without requiring allocation of a new trie.
+Keys do not need to be ordered in any specific way, but sorting them may make implementing the API easier.
+
+The following API endpoints provide all of this functionality:
+
+### Walk (Flat|Rec)
```
Comment (Johann150) [inline SPEC.md]: The `Flat` and `Rec` names are not very descriptive. Since we already have the trie/tree analogy, maybe these could be named something like `WalkTree` and `WalkLeaves` (adjusting the other functions appropriately).
For me, `Flat` does not really make sense here because that implies flattening the tree into a list or something. And I don't really get what `Rec` should stand for.

File: SPEC.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	var output_list
+	WalkRec (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
+```
+
+### Sub (Flat|Rec)
+SubFlat|SubRec takes a pattern, and constructs a new trie.
+The new trie will contain all of the key-value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.
```
Comment (Johann150) [inline SPEC.md]: Should this be `key-(raw)value` pairs?
```suggestion
The new trie will contain all of the key-(raw)value pairs that match under the corresponding `Walk*` function, but with the pattern (and the dot immediately after it) stripped.
```
I know it refers to another idea here, but still people might be confused.

File: SPEC.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	WalkFlat (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
+ListRec (pattern):
+	var output_list
+	WalkRec (pattern, (key, value){
+		output_list.append(value)
+	})
+	return output_list
```
Comment (Johann150) [inline SPEC.md]: The syntax of this pseudocode seems a bit strange to me, especially the anonymous function syntax. It's a bit difficult to convey an advanced concept like a closure in pseudocode, and I don't have a better idea (yet).

File: SPEC.md (hunk 9 above)
Comment (Johann150) [reply SPEC.md]: Also maybe removed instead of stripped? Stripped seems to me more of a slang word, while removed should be clearer.

File: SPEC.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	WalkFlat (pattern, (key, value){
+		output_trie[key.strip-from-start(pattern + ".")] = value
+	})
+	return output_trie
+SubRec (pattern):
+	var output_trie
+	WalkRec (pattern, (key, value){
+		output_trie[key.strip-from-start(pattern + ".")] = value
+	})
+	return output_trie
```
Comment (Johann150) [inline SPEC.md]: Same comment as with `ListFlat` and `ListRec` pseudocode applies here.

**Review by Johann150 — COMMENTED** [review]
Should we also recommend `KeyFlat` and `KeyReq` functions?

**Comment by Johann150** [comment]
And I forgot one more thing: For non Perl and/or `/[Rr]eg[Ee]xp?/` savvy people a list of code points for "vertical whitespace" would be nice, since vertical white space - unlike general whitespace - is not well defined by Unicode.
https://regex101.com says that `\v` in PCRE is identical to `[\x{2028}\n\r\x{000B}\f\x{2029}\x{0085}]`, so that would be these Unicode code points (with their C escape codes):

```
- U+000A Line Feed (`\n`)
- U+000B Vertical Tabulation (`\v`)
- U+000C Form Feed (`\f`)
- U+000D Carriage Return (`\r`)
- U+0085 Next Line
- U+2028 Line Separator
- U+2029 Paragraph Separator
```

File: SPEC.md (hunk 1 above)
Comment (CosmicToast) [reply SPEC.md:40]: The trailing non-vertical whitespace is not part of the value (as it may not end with whitespace).
This does not interact with flexspace (for the same reason).
If you want a value that starts/ends with whitespace, you must use raw values.

File: SPEC.md (hunk 3 above)
Comment (CosmicToast) [reply SPEC.md]: Will be added in draft 2.

File: SPEC.md (hunk 2 above)
Comment (CosmicToast) [reply SPEC.md]: I'll have to experiment with this in the grammar (right now a newline is the only terminator).
I agree that this makes sense though, so I'll try it and if it doesn't break anything, will change both.

File: SPEC.md (hunk 4 above)
Comment (CosmicToast) [reply SPEC.md]: LGTM, will be added in draft 2.

File: SPEC.md (hunk 5 above)
Comment (CosmicToast) [reply SPEC.md]: `#` has no special meaning inside of raw values, so I'll omit that part. Besides that, will be added in draft 2.

File: SPEC.md (hunk 6 above)
Comment (CosmicToast) [reply SPEC.md]: No, comments may not be placed within such constructs, since comments are statements and not whitespace.
This extension is already mostly there since it's harder to disallow it than it is to let it be.
Having comments inside of flexspace may be an extension, but I don't actually expect people to ever do this as it is.

File: SPEC.md (hunk 7 above)
Comment (CosmicToast) [reply SPEC.md]: You're right, core/bareword/03 will be removed with draft 2.

File: SPEC.md (hunk 8 above)
Comment (CosmicToast) [reply SPEC.md]: Rec was meant to stand for `Recursive`.
I'm ok with `Tree` and `Leaves`, will see if we get further comments re: the terminology.

File: SPEC.md (hunk 10 above)
Comment (CosmicToast) [reply SPEC.md]: Yeah, it'd be nice to have some conventions for pseudocode, but as of right now we don't have any and this is (hopefully) clear enough.

File: SPEC.md (hunk 9 above)
Comment (CosmicToast) [reply SPEC.md]: LGTM on both, will be changed in draft 2.

**Comment by CosmicToast** [comment]
P.S. I'm going to gather comments for a few days before writing/releasing draft 2, just to avoid having 500 total draft versions :)

File: SPEC.md (hunk 5 above)
Comment (Johann150) [reply SPEC.md]: I only put that there since I removed the "significant character" part. But if you think its okay like that it's also fine 😄

File: SPEC.md (hunk 10 above)
Comment (Johann150) [reply SPEC.md]: well we could look at what ~~Pseudocode~~ Python does
```suggestion
ListFlat (pattern):
	var output_list
	WalkFlat (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
ListRec (pattern):
	var output_list
	WalkRec (pattern, lambda key, value:
		output_list.append(value)
	)
	return output_list
```
looks a bit more concise to me, but this much 🤏

File: SPEC.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
```
Comment (Johann150) [inline SPEC.md]: I understand this as "fully optional for libraries to implement", while from #2 I gather it should mean fully optional to library implementers as well as library users.
```suggestion
All extensions are fully optional to be implemented. Implementations should allow for specific extensions to be en- or disabled.
Currently, multiple of them are consequences of keeping the grammar lean (so you will probably find it easier to implement them than not), and are documented, but only parts of their behaviors are desirable.
```

File: SPEC.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+* Between a (raw) value and a comment.
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
```
Comment (Johann150) [inline SPEC.md]: Does this mean that `core` does not allow empty lines? On the other hand, the name suggests this should only allow for arbitrary *horizontal* whitespace.
The lines between `core`, `flexspace` and `tabulation` are a bit blurry here.

File: SPEC.md (hunk 12 above)
Comment (CosmicToast) [reply SPEC.md]: When I say "users" I mean end-users (application users). I.e the people writing the CNI files.
Library users have the option on the basis of the library they end up picking.

The intended workflow is:
1. You realize you're working with a CNI file, you know what CNI is (`core`) and can write it.
2. If the application author tells you to use an extension (or you realize that an extension is available), you know you can take advantage of its features for that application only.

Basically, `ext` is for stuff that I don't want CNI as a whole to be associated with / preassumed availability of.
`ini` can in fact be considered to be `ext/ini`, but it's separated out (right now) for convenience purposes.

File: SPEC.md (hunk 13 above)
Comment (CosmicToast) [reply SPEC.md]: You're right, this should be *horizontal* whitespace.
`core` allows empty lines, and in fact core tests have empty lines (this is a mild reminder that the test suite and ref. impl is still the source of truth, SPEC.md is here to help people interpret the two of them).

**Comment by CosmicToast** [comment]
I think I went through everything here.
How's it looking now?

File: Parser.pm6 [hunk 14]
```diff
@@ -65,23 +65,23 @@ class pCNI is Hash {
...
 	}
 
 	# returns a list of all values directly in the section
-	method ListFlat($prefix) {
-		return self.SubFlat($prefix).values.list;
+	method ListLeaves($prefix) {
+		return self.SubLeaves($prefix).values.list;
 	}
 
 	# you are also recommended to implement WalkFlat and WalkRec alongside the above
```
Comment (Johann150) [inline Parser.pm6]:
```suggestion
	# you are also recommended to implement WalkLeaves and WalkTree alongside the above
```
I don't know if `KeyLeaves` and `KeyTree` might also be interesting.

File: SPEC.md [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+* Between a (raw) value and a comment.
+
+### Tabulation
+In implementations with the "tabulation" extension, whitespace before and after statements is not significant.
+This means you may have arbitrary horizontal whitespace in the following locations:
+* In front of a section declaration.
+* In front of a key-(raw)value pair.
+* In front of a comment.
+* Between a key-rawvalue pair and the next statement.
+* Between a section declaration and the next statement.
```
Comment (Johann150) [inline SPEC.md]: Couldn't this be simplified to "at the start of any line"?

File: SPEC.md [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+* Library authors may implement any amount of extensions they wish.
+* Application authors should only mention any extensions that are important to the functioning of their application.
+* Application end-users should not rely on any extensions being present, unless the application they use makes it clear that said extension is available.
+ 
+### Flexspace
+In implementations with the "flexspace" extension, whitespace within statements is not significant.
+This means that you may have arbitrary whitespace (not only spaces, but also tabs, newlines, zero-width spaces, and so on) in the following locations:
+* Between the opening section bracket, the section name, and the closing section bracket.
+* Between the key in a key-(raw)value pair, the `=`, and the (raw)value.
+* Between a (raw) value and a comment.
```
Comment (Johann150) [inline SPEC.md]: Horizontal whitespace between values and comments is already allowed by `core`, as are empty lines. Is this rule still necessary?

File: SPEC.md [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
```
Comment (Johann150) [inline SPEC.md]: I think to be consistent we should allow horizontal whitespace instead of just spaces.
```suggestion
It's a key, an equals, and a value, with arbitrary amounts of horizontal whitespace before and after the `=`.
```

**Review by Johann150 — COMMENTED** [review]
I hope you'll not beat me with a stick for this since you are pushing for the Raku grammar to be the sole specification, but I made some EBNF syntax files for CNI. They helped in noticing some odd things to annoy you with. :P
https://github.com/Johann150/cni-format/tree/ebnf/ebnf

File: Parser.pm6 (hunk 14 above)
Comment (CosmicToast) [reply Parser.pm6]: Walker pattern gets called walk, surely.
It's not like it gives you a set of keys, eh?

File: SPEC.md (hunk 15 above)
Comment (CosmicToast) [reply SPEC.md]: The last two translate to something beyond that. The first 3 could, though, I'll think about it.

File: SPEC.md (hunk 16 above)
Comment (CosmicToast) [reply SPEC.md]: I think it's fine to keep it for clarity, since this is a companion document.

File: SPEC.md (hunk 17 above)
Comment (CosmicToast) [reply SPEC.md]: I think core should only require people to check for ' ' as spacings, to simplify some parsers.
So if anything I would rather reverse it (if they do non-significant whitespace properly they end up auto-implementing the flexspace and tabulation extensions, eh?)
OTOH, users might want to be able to insert hard tabs.
I'll have to think about this.
Might end up removing tabulation/flexspace and adding those into core tbh.

**Comment by CosmicToast** [comment]
I'll think about all of these and do a draft 3 this weekend.
Re: EBNF - I don't mind it existing, I just don't want it to be "official".
I encourage you to add it to the project's wiki though (with the obvious disclaimers)!

File: Parser.pm6 (hunk 14 above)
Comment (Johann150) [reply Parser.pm6]: Yes, the walk functions stay as they are (apart from updating their names to be consistent as in my code suggestion). I meant KeyLeaves/KeyTree as *additional* API functions might be interesting, since we also have the List... functions.

Comment (CosmicToast) [reply Parser.pm6]: Oh, now I see.
I think that's interesting, as well as SectionTree and SectionLeaves.
I don't want to require too much from library writers though.
Then again, the API is purely a suggestion / informative...

**Comment by Johann150** [comment]
If you mean the literal wiki, that won't be possible since I do not have permissions to edit the wiki. I would rather make a pull request to add the files into the repository after we are finished here and I've updated the files accordingly.

**Comment by CosmicToast** [comment]
I do mean the literal wiki.
Looks like it had some restrictions on it - I've removed them.
Like I said, I don't want it to be official, and I consider everything into the repo proper to be official, thus the conundrum :slightly_smiling_face: 
I'm fine with as many informative/unofficial things as people want, just not the repo proper - and that's exactly what the attached wikis are for.

**Comment by CosmicToast** [comment]
I've decided to integrate flexspace and tabulation into core cni.
The reason I've excluded them so far is because they allow users to do error prone things.
For example:
```
[
    this.is.valid
]
[ # this is not valid
     neither # is this
] # but this is
```

Instead, I'm simply going to have a mention in the SPEC that some technically valid syntax is discouraged, and point people to the testing suite.

Either way, this means I need to rewrite a good chunk of the tests, so this is why things are taking a while, sorry ^^;;

**Comment by CosmicToast** [comment]
Sorry this took a bit.
It's a relatively large-ish update, and real life has been getting in the way of work recently.
It still is, mind you, but hopefully with smaller future iterations, more can get done more often.

File: SPEC.md [hunk 18]
```diff
@@ -132,10 +128,25 @@ SubTree (pattern):
...
+### Section (Leaves|Tree)
+SectionLeaves|SectionTree takes a pattern and returns a set of prefixes.
+It can be calculated roughly as follows:
+1. Get the corresponding list of KeyLeaves|KeyTree.
+2. For each key:
+	1. If the key does not contain a `.`, continue.
+	2. Remove the final `.` and all text that follows from the key.
+	3. Add the result to the result set.
+	4. Add the result to the list of keys to process.
+3. Return the resulting set.
```
Comment (Johann150) [inline SPEC.md]: I'm not sure my alternate phrasing for 2.2 makes it clearer or more confusing, but here it goes:

```suggestion
1. Get the corresponding set from KeyLeaves|KeyTree.
2. For each key in the set:
	1. If the key does not contain a `.`, continue.
	2. Remove the last section of the key (i.e. what would be matched by the PCRE `/\.[^.]+$/m`).
	3. Add the result to the result set.
	4. Add the result to the set of keys to process.
3. Return the resulting set.
````

**Review by Johann150 — APPROVED** [review]

File: SPEC.md (hunk 18 above)
Comment (CosmicToast) [reply SPEC.md]: You're right re: list -> set.
Re: 2.2, I think a better approach would be to simply give an example.
Will try and do that soon.
Also, should probably make KeyLeaves|Tree into a set as well.

**Comment by CosmicToast** [comment]
leaving the branch up in case something comes up that required quickly reverting and working on it more
